### PR TITLE
Add a Lint for Pointer to Integer Transmutes in Consts

### LINKED ISSUE #87525: Add lint against ptr-to-int transmutes in consts
We do not allow ptr-to-int casts in `const` context, so people might be tempted to do a transmute instead. This will not work, and if they ever actually test their code during CTFE they will get an error, but it might still be good to add an extra layer of defense here (since this is a rather odd corner of the language) and have a lint for this.

@jhpratt started working on this, [here's a link](https://github.com/rust-lang/rust/commit/6351c670f654ead12ec64534fcc64cb32dc380c8).

### PR DESCRIPTION
**Pull request by veera-sivarajan**
Fixes #87525 

This PR adds a MirLint for pointer to integer transmutes in const functions and associated consts. The implementation closely follows this comment: https://github.com/rust-lang/rust/pull/85769#issuecomment-880969112. More details about the implementation can be found in the comments.

Note: This could break some sound code as mentioned by RalfJung in https://github.com/rust-lang/rust/pull/85769#issuecomment-886491680:

> ... technically const-code could transmute/cast an int to a ptr and then transmute it back and that would be correct -- so the lint will deny some sound code. Does not seem terribly likely though.

References:
1. https://doc.rust-lang.org/std/mem/fn.transmute.html
2. https://doc.rust-lang.org/reference/items/associated-items.html#associated-constants

### CODE REVIEW

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -5095,3 +5096,37 @@ declare_lint! {
...
+    /// But sometimes the compiler might not emit an error for pointer to integer transmutes
+    /// inside const functions and associated consts because they are evaluated only when referenced.
+    /// Therefore, this lint serves as an extra layer of defense to prevent any undefined behavior
+    /// from compiling without any warnings or errors.
+    ///
+    /// See [std::mem::transmute] in the reference for more details.
+    ///
+    /// [std::mem::transmute]: https://doc.rust-lang.org/std/mem/fn.transmute.html
+    pub PTR_TO_INTEGER_TRANSMUTE_IN_CONSTS,
+    Deny,
```
Comment (veera-sivarajan) [inline compiler/rustc_lint_defs/src/builtin.rs:5130]: Not sure if this should be `Deny` or `Warn` by default.

**Comment by veera-sivarajan** [comment]
Transmuting a pointer to integer in const context is undefined behavior[^1].

Also, this lint specifically targets const functions and associated consts because they are evaluated only when referenced. This can result in undefined behavior in a library going unnoticed until the function or constant is actually used.

Moreover, this lint will not affect the functions you have linked because:
1. `addr()` is not a const function.
2. `without_provenance()` transmutes an integer to pointer.


[^1]: https://doc.rust-lang.org/std/mem/fn.transmute.html#transmutation-between-pointers-and-integers

**Comment by RalfJung** [comment]
> I'd be tempted to run it on optimized mir where we can have const-folded a bunch of those cases away already, but then it'd be super inconsistent which we probably don't want either :/

Also, we're not optimizing const MIR.

**Comment by RalfJung** [comment]
> Seem hard to lint on when it's not always UB, because things like NonNull::dangling().addr() is a transmute but isn't UB.

Indeed, that's the tricky part. The lint triggers in this code which is exactly an example of that:

https://github.com/rust-lang/rust/blob/b5bd0fe48a0427a82804759a5acd795851c65e4b/library/core/src/ptr/mod.rs#L1981-L1985

File: compiler/rustc_lint_defs/src/builtin.rs (hunk 1 above)
Comment (estebank) [reply compiler/rustc_lint_defs/src/builtin.rs:5130]: I guess that warn by default is a reasonable default to land this with, and in a few releases, once we have more confidence that we don't have any false positives you can increase it to deny.

File: src/tools/clippy/tests/ui/pointer_to_integer_transmutes_in_consts.rs
[the remark this answers is not in the document]
Comment (estebank) [reply src/tools/clippy/tests/ui/pointer_to_integer_transmutes_in_consts.rs:1]: Please address this request.

File: library/core/src/ptr/mod.rs [hunk 2]
```diff
@@ -1916,6 +1916,7 @@ pub unsafe fn write_volatile<T>(dst: *mut T, src: T) {
 /// than trying to adapt this to accommodate that change.
 ///
 /// Any questions go to @nagisa.
+#[cfg_attr(not(bootstrap), allow(ptr_to_integer_transmute_in_consts))]
```
Comment (estebank) [inline library/core/src/ptr/mod.rs:1919]: Given the description of the lint, isn't the comment at https://github.com/rust-lang/rust/pull/130540/files#diff-c175d4e27676febf62c061d31cf9756d256b46e2e44cc6b3177d4ff75e932567R1982-R1986 which causes this to be allowed here incorrect?

**Review by estebank — COMMENTED** [review]
The code change looks correct, but I have some questions.

Shouldn't we also account for `as`? Is `align_offset` invoking UB?

**Comment by RalfJung** [comment]
> Shouldn't we also account for as? 

`as` from ptr to int is just forbidden in `const` contexts.

> Is align_offset invoking UB?

No, that function is treated specially by the const interpreter to ensure that the pointer argument is always actually an integer.

File: src/tools/clippy/tests/ui/pointer_to_integer_transmutes_in_consts.rs
[the remark this answers is not in the document]
Comment (veera-sivarajan) [reply src/tools/clippy/tests/ui/pointer_to_integer_transmutes_in_consts.rs:1]: Thank you for the detailed feedback.

**Pull request merged**